### PR TITLE
Scanner now works

### PR DESCRIPTION
**Pull request by KitchenTable99**
* Added test cases to HelloWorld.java for string edge cases, comments, and math stuff
* changed Main.java to prevent double printing of EOF token
* added token type logic to Scanner.java
* fixed complete string, complete math, and compelte comment methods
* identified a potential bug

**next steps:** figure out how to handle error s and start handling the errors described in the directions

### CODE REVIEW

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 1]
```diff
@@ -186,26 +152,34 @@ private boolean isCompleteComment(Stack<Character> spellingStack) {
...
     }
 
     private boolean isCompleteMath(Stack<Character> spellingStack) {
         char lastChar = spellingStack.peek();
-        if (lastChar == ' ' || lastChar == '=') {
-            // set token type
+        if (Character.isWhitespace(lastChar)) {
+            skippedLastToken = spellingStack.pop();
+            return true;
+        } else if (lastChar == '=') {
```
Comment (KitchenTable99) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java]: All the branches that just reset skippedLastToken and return true could be combined

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 2]
```diff
@@ -260,9 +234,15 @@ private boolean isCompleteInt(Stack<Character> spellingStack) {
...
-            return spellingStack.peek() == '"' && spellingStack.get(stackSize - 1) != '\\';
+            if (spellingStack.peek() == '"' && spellingStack.get(stackSize - 2) != '\\') {
+                skippedLastToken = '\0';
+                return true;
+            } else {
+                return false;
+            }
         } else if (stackSize > 5000) {
             //raise error;
+            skippedLastToken = '\0';
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:246]: Maybe instead of checking the string length here we can throw in a call to a boolean "checkStringErrors()" method in the getTokenKind method. This would be where we register the string length and multiline errors. if it returns false then create a string token, otherwise create an error token. This is also helpful because prematurely ending strings based on length could create a bunch of problems in future token scanning / checking for unterminated strings.

Im not sure what the elegant way of checking for illegal escape characters in strings would be. It would make sense to register them each one is scanned but then the string error checking would be separated. Maybe we need a "stringContainsErrors" field to register if the current token contains errors and then that becomes a case in the "checkStringErrors()" method? some thoughts for the next branch

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 3]
```diff
@@ -289,132 +269,86 @@ private boolean isCompleteIdentifier(Stack<Character> spellingStack) {
...
+                    return Kind.UNARYDECR;
+                default:
+                    return null;
+            }
+        } else if (leadingChar == '=' && spellingStack.size() == 1) {
+            return Kind.ASSIGN;
+        } else if (leadingChar == '=') {
+            return Kind.COMPARE;
+        } else if (Character.isDigit(leadingChar)) {
+            return Kind.INTCONST;
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:335]: We can check Int size here if invalid return Kind.ERROR;

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 4]
```diff
@@ -289,132 +269,86 @@ private boolean isCompleteIdentifier(Stack<Character> spellingStack) {
...
+        } else if (leadingChar == '"') {
+            return Kind.STRCONST;
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:337]: This is where I was thinking about adding a call to "checkStringErrors()"

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 5]
```diff
@@ -289,132 +269,86 @@ private boolean isCompleteIdentifier(Stack<Character> spellingStack) {
...
+        } else {
+            return Kind.IDENTIFIER; // todo: assumes same as check token with respect to else clause.
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:339]: Default Token type should be an error and catch if there are illegal characters scanned. Can add a check to see if leadingChar is alphabetical and currentTokenLegal is true.

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 6]
```diff
@@ -108,11 +80,11 @@ public Scanner(Reader reader, ErrorHandler handler) {
...
-        if (skippedLastToken != '\0') {
+        if (skippedLastToken != '\0' && !Character.isWhitespace(skippedLastToken)) {
             spellingStack.push(skippedLastToken);
         }
 
-        do {
+        while (!isCompleteToken(spellingStack)) {
             try {
                 char letter = sourceFile.getNextChar();
                 if (!Character.isWhitespace(letter) || !spellingStack.empty()) {
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:90]: I assume it would not be good to add a flag here that finds if a character is illegal ?

If its a good enough reason to modify scan I would check if the current token is a string or comment and whether the char is legal. If the char is not legal and the current token is not a string or comment, register an error and set a "currentTokenLegal" field / flag to false. Unsupported escape characters within a string would be handled separately.


Lastly, It could be useful to store the lineNumber for the start of each token

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 7]
```diff
@@ -122,17 +94,10 @@ public Token scan() {
                 // if there are no more character then check to see if the final token is invalid
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:95]: Also check unterminated strings and block comments

**Review by ejcohe22 — APPROVED** [review]
I like your changes. I dont know if there is a non-astonishing way to separate out the 
"
skippedLastToken = '\0';
return true;
"
block, but I would agree that it is a piece of code that is repeated often and could be simplified. I was thinking a good name would be "private boolean handleEndOfToken()". The method simply always returns true. Im concerned it would be confused with the functionality of createToken...

I included some thoughts for handling errors. If you're good with everything we can move this up and I can start working on a new error branch!

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java [hunk 8]
```diff
@@ -186,26 +152,34 @@ private boolean isCompleteComment(Stack<Character> spellingStack) {
         char secondChar = spellingStack.get(1);
         char lastChar = spellingStack.peek();
         if (secondChar == '/') {
-            return lastChar == '\n' || lastChar == '\r';
+            if (lastChar == '\n' || lastChar == '\r') {
+                skippedLastToken = '\0';
```
Comment (ejcohe22) [inline src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java]: Should this be popped from the comment before creating the token (dont include newlines in single line comment)?

File: src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java (hunk 2 above)
Comment (ejcohe22) [reply src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:245]: Looking at the instructions I am more confident in saying illegal string tokens should be fully scanned and registered as an error token. This will also prevent ruining the next token and make it possible to find unterminated strings.

Comment (KitchenTable99) [reply src/proj7BittingCerratoCohenEllmer/bantam/lexer/Scanner.java:246]: String constants cannot span multiple lines in Bantam java though I agree with the premature ending for sure. Illegal escape characters is \ followed by anything but n,t,",\,f.
